### ISSUE TITLE
Incorrect coordinates in swaymsg get_tree

### ISSUE THREAD

**Issue by schauveau** [swaywm/sway#3821]
This is a followup on issue #3785 which was IMHO closed too early.
The issue is that the coordinates given in `rect`, `window_rect` and `deco_rect` are not consistent with i3. For convenience, I will refer to the current window as WIN and to its parent container as PARENT. 

In i3:
   - `WIN.rect` provides the **absolute** window coordinates **excluding** the decoration but including the borders.
   - `WIN.window_rect` is **relative** to `WIN.rect` and provides the coordinates excluding the borders and the decorations.
   - `WIN.deco_rect` is **relative** to `PARENT.rect` and provides the coordinates of the decoration for WIN only.   

In Sway:
  - `WIN.rect` provides the **absolute** window coordinates **INCLUDING THE DECORATIONS** and the borders.
  - `WIN.window_rect` has the same value than in i3 which is useless since that makes it relative to neither `WIN.rect` nor `PARENT.rect`. 
  - `WIN.deco_rect` is **relative** to `WIN.rect`. Also, in stacked or tabbed mode, it provides the combined coordinates of all decorations in the PARENT container instead of the individual decoration as in i3.

**Comment by schauveau** [comment]
I also noticed a few other differences that may affect the future implementation.
- in i3 each floating window is described by 2 nodes (the actual window and its PARENT container) but in sway only one node is used (so no usable PARENT).
- in i3 a window with a border of type `none` or `pixel` has a usable `deco_rect` when its parent layout is `stack` or `tabbed`. That makes sense since those containers are drawing a decoration for each of their elements. In Sway, the `deco_rect` is incorrectly filled with zeros.

**Comment by RedSoxFan** [comment]
> This is a followup on issue #3785 which was IMHO closed too early.
>> Am I right to assume that window_rect is supposed to provide the relative geometry of the window without its decorations?

#3785 was closed because the question quoted above was answered and better documented. Since #3785 was focused on `window_rect`, this is technically a separate issue (`deco_rect` and `rect`) anyway. Sorry, I didn't get to your follow-up response sooner though

As for this issue, #3823 should correct both `rect` and `deco_rect`, but please let me know if you notice a case that I missed.

> in i3 each floating window is described by 2 nodes (the actual window and its PARENT container) but in sway only one node is used (so no usable PARENT).

I think that sway's current response is sufficient (or at least will be after #3823) in regard to this. The parent is the workspace. @ddevault @emersion thoughts?

**Comment by emersion** [comment]
>The parent is the workspace.

I can see how this could break scripts, however the Sway description seems more accurate (doesn't leak internal implementation details). I don't really know whether changing this is a big deal or not.

**Comment by schauveau** [comment]
Thanks. This patch is now making window_rect useful for example to take screenshots of the current window. I tested a few windows and so far `rect` and `window_rect` are working as expected except in the case of a stacked container where the area described by `rect` still contains all the decorations except the first. Simply speaking, for a stacked parent container, the code at ipc-json.c:537 should consider the height of the decoration multiplied by the number of siblings. 
I did not test deco_rect yet.

**Comment by schauveau** [comment]
So far all my tests on deco_rect are successful including on floating windows.

**Comment by schauveau** [comment]
There is a problem in fullscreen mode where a decoration is incorrectly reported.
Current:    
```
  "rect":               {  "x": 0, "y": 27, "width": 1920, "height": 1053 },
  "window_rect": {  "x": 0, "y": 0, "width": 1920, "height": 1080 },
  "deco_rect":     {  "x": -1280, "y": 0, "width": 1920, "height": 27 },
```
Expected:
```
  "rect":               { "x": 0, "y": 0, "width": 1920, "height": 1080 },
  "window_rect": { "x": 0, "y": 0, "width": 1920, "height": 1080 },
  "deco_rect":     { "x": 0, "y": 0, "width": 0, "height": 0 },
```

**Comment by RedSoxFan** [comment]
> Simply speaking, for a stacked parent container, the code at ipc-json.c:537 should consider the height of the decoration multiplied by the number of siblings.

#3828 

> There is a problem in fullscreen mode where a decoration is incorrectly reported.

#3827

**Comment by RedSoxFan** [comment]
@schauveau, I think we are going to hold off on changing the way floating containers are represented until a sound argument is made for making the change, but all of your other points have been fixed and this can be closed, correct?

**Comment by schauveau** [comment]
I agree. Having a parent via `floating_nodes` instead of `nodes` is probably not a big deal for most use cases. As far as I can tell, everything is working as expected so I think that the whole issue can be closed. Thanks
